tools/src/wyvern/target/corewyvernIL/astvisitor/TailCallVisitor.java

```diff
@@ -0,0 +1,240 @@
+package wyvern.target.corewyvernIL.astvisitor;
+
+import wyvern.target.corewyvernIL.ASTNode;
+import wyvern.target.corewyvernIL.Case;
+import wyvern.target.corewyvernIL.Environment;
+import wyvern.target.corewyvernIL.FormalArg;
+import wyvern.target.corewyvernIL.binding.Binding;
+import wyvern.target.corewyvernIL.decl.Declaration;
+import wyvern.target.corewyvernIL.decl.DefDeclaration;
+import wyvern.target.corewyvernIL.decl.DelegateDeclaration;
+import wyvern.target.corewyvernIL.decl.TypeDeclaration;
+import wyvern.target.corewyvernIL.decl.ValDeclaration;
+import wyvern.target.corewyvernIL.decl.VarDeclaration;
+import wyvern.target.corewyvernIL.decltype.AbstractTypeMember;
+import wyvern.target.corewyvernIL.decltype.ConcreteTypeMember;
+import wyvern.target.corewyvernIL.decltype.DeclType;
+import wyvern.target.corewyvernIL.decltype.DefDeclType;
+import wyvern.target.corewyvernIL.decltype.ValDeclType;
+import wyvern.target.corewyvernIL.decltype.VarDeclType;
+import wyvern.target.corewyvernIL.expression.Bind;
+import wyvern.target.corewyvernIL.expression.Cast;
+import wyvern.target.corewyvernIL.expression.Expression;
+import wyvern.target.corewyvernIL.expression.FFIImport;
+import wyvern.target.corewyvernIL.expression.FieldGet;
+import wyvern.target.corewyvernIL.expression.FieldSet;
+import wyvern.target.corewyvernIL.expression.IExpr;
+import wyvern.target.corewyvernIL.expression.IntegerLiteral;
+import wyvern.target.corewyvernIL.expression.Let;
+import wyvern.target.corewyvernIL.expression.Match;
+import wyvern.target.corewyvernIL.expression.MethodCall;
+import wyvern.target.corewyvernIL.expression.New;
+import wyvern.target.corewyvernIL.expression.Path;
+import wyvern.target.corewyvernIL.expression.RationalLiteral;
+import wyvern.target.corewyvernIL.expression.StringLiteral;
+import wyvern.target.corewyvernIL.expression.Variable;
+import wyvern.target.corewyvernIL.metadata.IsTailCall;
+import wyvern.target.corewyvernIL.support.EmptyTypeContext;
+import wyvern.target.corewyvernIL.support.TypeContext;
+import wyvern.target.corewyvernIL.support.Util;
+import wyvern.target.corewyvernIL.type.CaseType;
+import wyvern.target.corewyvernIL.type.DataType;
+import wyvern.target.corewyvernIL.type.ExtensibleTagType;
+import wyvern.target.corewyvernIL.type.NominalType;
+import wyvern.target.corewyvernIL.type.StructuralType;
+import wyvern.target.corewyvernIL.type.ValueType;
+
+/** TailCallVisitor finds all MethodCall expressions in a
+ *  corewyvernIL AST and attaches an IsTailCall metadata to them
+ *  if the method calls are tail calls.
+ */
+public class TailCallVisitor extends ASTVisitor<Boolean, Void> {
+
+    public static void annotate(IExpr program) {
+        program.acceptVisitor(new TailCallVisitor(), false);
+    }
+
+    public Void visit(Boolean inTailPosition, New newExpr) {
+        for (Declaration decl : newExpr.getDecls()) {
+            decl.acceptVisitor(this, inTailPosition);
```

Comment:
This doesn't really matter because Val- and Var-Declaration pass "false" for inTailPosition, but it might be better conceptually to also pass "false" here.